rust/src/sip/detect.rs

```diff
@@ -80,88 +126,662 @@ pub unsafe extern "C" fn rs_sip_tx_get_protocol(
         }
     }
-
     *buffer = ptr::null();
     *buffer_len = 0;
+    return false;
+}
 
+unsafe extern "C" fn sip_request_line_setup(
+    de: *mut c_void, s: *mut c_void, _raw: *const std::os::raw::c_char,
+) -> c_int {
+    if DetectSignatureSetAppProto(s, ALPROTO_SIP) != 0 {
+        return -1;
+    }
+    if DetectBufferSetActiveList(de, s, G_SIP_REQUEST_LINE_BUFFER_ID) < 0 {
+        return -1;
+    }
     return 0;
 }
 
-#[no_mangle]
-pub unsafe extern "C" fn rs_sip_tx_get_request_line(
-    tx: &mut SIPTransaction, buffer: *mut *const u8, buffer_len: *mut u32,
-) -> u8 {
+unsafe extern "C" fn sip_request_line_get(
+    de: *mut c_void, transforms: *const c_void, flow: *const c_void, flow_flags: u8,
+    tx: *const c_void, list_id: c_int,
+) -> *mut c_void {
+    return DetectHelperGetData(
+        de,
+        transforms,
+        flow,
+        flow_flags,
+        tx,
+        list_id,
+        sip_request_line_get_data,
+    );
+}
+
+unsafe extern "C" fn sip_request_line_get_data(
+    tx: *const c_void, _flags: u8, buffer: *mut *const u8, buffer_len: *mut u32,
+) -> bool {
+    let tx = cast_pointer!(tx, SIPTransaction);
     if let Some(ref r) = tx.request_line {
         if !r.is_empty() {
             *buffer = r.as_ptr();
             *buffer_len = r.len() as u32;
-            return 1;
+            return true;
         }
     }
-
     *buffer = ptr::null();
     *buffer_len = 0;
+    return false;
+}
 
+unsafe extern "C" fn sip_response_line_setup(
+    de: *mut c_void, s: *mut c_void, _raw: *const std::os::raw::c_char,
+) -> c_int {
+    if DetectSignatureSetAppProto(s, ALPROTO_SIP) != 0 {
+        return -1;
+    }
+    if DetectBufferSetActiveList(de, s, G_SIP_RESPONSE_LINE_BUFFER_ID) < 0 {
+        return -1;
+    }
     return 0;
 }
 
-#[no_mangle]
-pub unsafe extern "C" fn rs_sip_tx_get_response_line(
-    tx: &mut SIPTransaction, buffer: *mut *const u8, buffer_len: *mut u32,
-) -> u8 {
+unsafe extern "C" fn sip_response_line_get(
+    de: *mut c_void, transforms: *const c_void, flow: *const c_void, flow_flags: u8,
+    tx: *const c_void, list_id: c_int,
+) -> *mut c_void {
+    return DetectHelperGetData(
+        de,
+        transforms,
+        flow,
+        flow_flags,
+        tx,
+        list_id,
+        sip_response_line_get_data,
+    );
+}
+
+unsafe extern "C" fn sip_response_line_get_data(
+    tx: *const c_void, _flags: u8, buffer: *mut *const u8, buffer_len: *mut u32,
+) -> bool {
+    let tx = cast_pointer!(tx, SIPTransaction);
     if let Some(ref r) = tx.response_line {
         if !r.is_empty() {
             *buffer = r.as_ptr();
             *buffer_len = r.len() as u32;
-            return 1;
+            return true;
         }
     }
+    *buffer = ptr::null();
+    *buffer_len = 0;
+    return false;
+}
 
+unsafe extern "C" fn sip_get_header_value(
+    tx: *const c_void, i: u32, direction: u8, buffer: *mut *const u8, buffer_len: *mut u32,
+    strname: *const c_char,
```

Comment:
`strname` should use a pure rust type instead of converting back and forth

Reply:
As I reported in the previous PR (https://github.com/OISF/suricata/pull/11620#discussion_r1734995838):
```
The compiler complains about that:
extern fn uses type `str`, which is not FFI-safe

Same for string.
```

Reply:
Yes, this function `sip_get_header_value should not be labelled `extern` :

You can squash this commit in 
https://github.com/OISF/suricata/commit/6b7098d2f7d83c88b2369d86a00a911466bdb572